ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/shared/target/RepositoryBundleContainerAdapterFactory.java

```diff
@@ -0,0 +1,222 @@
+				repositoryImage = null;
+			}
+		}
+	};
+
+	private static final ITargetLocationHandler LOCATION_HANDLER = new ITargetLocationHandler() {
+
+		@Override
+		public IStatus reload(ITargetDefinition target, ITargetLocation[] targetLocations, IProgressMonitor monitor) {
+			for (ITargetLocation location : targetLocations) {
+				if (location instanceof RepositoryBundleContainer container) {
+					container.reload();
+				}
+			}
+			return Status.OK_STATUS;
+		}
+
+		@Override
+		public IWizard getEditWizard(ITargetDefinition target, TreePath treePath) {
+			Object segment = treePath.getLastSegment();
+			if (segment instanceof RequirementNode node) {
+				// TODO maybe we can support a requirements editor?
+				segment = node.container;
+			}
+			if (segment instanceof RepositoryBundleContainer container) {
+				RepositoryLocationWizard wizard = new RepositoryLocationWizard();
+				wizard.setTarget(target);
+				wizard.setBundleContainer(container);
+				return wizard;
+			}
+			return null;
+		}
+
+		@Override
+		public boolean canEdit(ITargetDefinition target, TreePath treePath) {
+			Object segment = treePath.getLastSegment();
+			return segment instanceof RepositoryBundleContainer || segment instanceof RequirementNode;
+		}
+
+		public boolean canRemove(ITargetDefinition target, TreePath treePath) {
+			return treePath.getLastSegment() instanceof RequirementNode;
+		}
+
+		public IStatus remove(ITargetDefinition target, TreePath[] treePaths) {
+			boolean reload = false;
+			for (TreePath path : treePaths) {
+				Object lastSegment = path.getLastSegment();
+				if (lastSegment instanceof RequirementNode node) {
+					RepositoryBundleContainer container = node.container;
+					RepositoryBundleContainer newContainer = new RepositoryBundleContainer(container.getUri(),
+							container.getRequirements().stream().filter(req -> req != node.requirement).toList());
+					ITargetLocation[] targetLocations = target.getTargetLocations();
+					for (int i = 0; i < targetLocations.length; i++) {
+						ITargetLocation loc = targetLocations[i];
+						if (loc == container) {
+							targetLocations[i] = newContainer;
+						}
+
+					}
+				}
+			}
+			return reload
+					? new Status(IStatus.OK, IPDEConstants.UI_PLUGIN_ID, ITargetLocationHandler.STATUS_FORCE_RELOAD,
+							"reloaded", null) //$NON-NLS-1$
+					: Status.OK_STATUS;
+		}
+	};
+
+	private static final ITreeContentProvider TREE_CONTENT_PROVIDER = new ITreeContentProvider() {
+
+		@Override
+		public boolean hasChildren(Object element) {
+			if (element instanceof RepositoryBundleContainer container) {
+				return !container.getRequirements().isEmpty();
+			}
+			return false;
+		}
+
+		@Override
+		public Object getParent(Object element) {
+			if (element instanceof RequirementNode node) {
+				return node.container;
+			}
+			return null;
+		}
+
+		@Override
+		public Object[] getElements(Object inputElement) {
+			return EMPTY_OBJECTS; // will never be called...
```

Comment:
Might as well throw an exception if it's never called.

Reply:
I think throwing an exception is not really required as it is not forbidden to call the method but PDE currently do not do this, and handling it graceful seems okay here.